cornell-dti/course-plan
# Add analytics data to CoursePlan route

### Summary 

This pull request adds the most recent TrackUsers analytics data directly to a new route on CoursePlan (courseplan.io/analytics) designed similarly to the 404 page. This should allow anyone on CP to see the daily prod runs (#601) without needing permissions to the actual prod Firestore.

- [x] Created an analytics route (`/analytics`)
- [x] Pretty print most recent TrackUsers document
- [x] Add a timestamp so people know when it was last run
- [x] Hide data (and style route differently) for those not on CP
- [x] Add timestamp field to documents, optimize code to only retrieve doc with newest timestamp 

![image](https://user-images.githubusercontent.com/25535093/143977345-3472c6a1-6f11-413c-be26-cdf604bc322b.png)

Remaining TODOs:

- [x] Add the `CPWhitelist` collection to prod and update it so all current team members are on it
- [x] Add the `CPWhitelist` rule to prod (see below) so data can only be read by those on the whitelist

![image](https://user-images.githubusercontent.com/25535093/143977228-2f6918cf-fc13-4aa1-8413-dd2aea370703.png)

### Test Plan 

1. Ensure that the most recent data run is shown with a timestamp
2. Confirm the /analytics route cannot be accessed unless logged in.
3. Confirm route does not show Firestore information when logged in to an email of someone not on `CPWhitelist` (and is nicely styled).

![image](https://user-images.githubusercontent.com/25535093/143977181-62c81fe3-1567-4d16-9604-46d37fa8ff09.png)

### Notes 

I unsuccessfully tried to block users not on the CPWhitelist from accessing the new `/analytics` route all together. If that would be significantly more desirable, let me know and I can further investigate this option! But the Firestore rules should prevent anyone who should not have access from having it.
